libs/execution_context/ebpf_link.c

```diff
@@ -89,12 +85,97 @@ C_ASSERT(
     EBPF_OFFSET_OF(ebpf_extension_dispatch_table_t, function) ==
     EBPF_OFFSET_OF(ebpf_extension_program_dispatch_table_t, ebpf_program_invoke_function));
 
+NTSTATUS
+_ebpf_link_client_attach_provider(
+    HANDLE nmr_binding_handle,
+    _Inout_ void* client_context,
+    _In_ const NPI_REGISTRATION_INSTANCE* provider_registration_instance)
+{
+    NTSTATUS status;
+    ebpf_link_t* link = (ebpf_link_t*)client_context;
+    void* provider_binding_context;
+    void* provider_dispatch;
+    const ebpf_extension_data_t* provider_data =
+        (const ebpf_extension_data_t*)provider_registration_instance->NpiSpecificCharacteristics;
+
+    bool lock_held = false;
+
+    ebpf_lock_state_t state = ebpf_lock_lock(&link->lock);
+    lock_held = true;
+
+    UNREFERENCED_PARAMETER(nmr_binding_handle);
+
+    // Verify that that the provider is using the same version of the extension as the client.
+    if (provider_data->version > EBPF_ATTACH_PROVIDER_DATA_VERSION ||
+        provider_data->size < sizeof(ebpf_extension_data_t)) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    ebpf_attach_provider_data_t* attach_provider_data = (ebpf_attach_provider_data_t*)provider_data->data;
+
+    if (provider_registration_instance->ModuleId->Type != MIT_GUID) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    if (memcmp(&provider_registration_instance->ModuleId->Guid, &link->attach_type, sizeof(link->attach_type)) != 0) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    if (memcmp(&attach_provider_data->supported_program_type, &link->program_type, sizeof(link->program_type)) != 0) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    // Only one provider can be attached to a link.
+    if (link->provider_attached) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    link->link_type = attach_provider_data->link_type;
+    link->bpf_attach_type = attach_provider_data->bpf_attach_type;
+
+    ebpf_lock_unlock(&link->lock, state);
+    lock_held = false;
+
+    status = NmrClientAttachProvider(
+        nmr_binding_handle, link, &_ebpf_link_dispatch_table, &provider_binding_context, &provider_dispatch);
+
+    if (!NT_SUCCESS(status)) {
+        goto Done;
+    }
+
+    state = ebpf_lock_lock(&link->lock);
+    lock_held = true;
+
+    link->provider_attached = true;
+
+Done:
+    if (lock_held)
+        ebpf_lock_unlock(&link->lock, state);
+
+    return status;
+}
+
+NTSTATUS
+_ebpf_link_client_detach_provider(void* client_binding_context)
+{
+    ebpf_link_t* link = (ebpf_link_t*)client_binding_context;
+    ebpf_lock_state_t state = ebpf_lock_lock(&link->lock);
+    link->provider_attached = false;
```

Comment:
is the link state not updated in this case? Should there be any safeguard against a buggy hook provider detaching when there are ongonig calls?